Streamlit으로 ui 구성하기

ysooj opened the pull request:
Streamlit을 활용해서 챗봇의 ui를 구성합니다.
- set_page_config : 브라우저 탭의 내용을 구성하고, layout과 sidebar 옵션을 선택합니다.
- st.markdown : 챗봇 이름을 꾸밀 수 있습니다.
- 간단한 채팅을 할 수 있는 챗봇을 만들었습니다.

ysooj committed:
(add, 머신러닝 데이터셋 불러오기)
토글 부분은 성진님 코드 참고했습니다.

ysooj committed:
전처리 작업 추가
불필요한 문구(예: [스파르타코딩클럽], 저작권 문구, 모든 토글 닫기 단축키) 제거
불필요한 공백, 특수 문자, 줄바꿈 제거

ysooj committed:
전처리 후 txt 파일로 로컬 폴더(cleaned_texts)에 저장하기
폴더가 없으면 만들고, txt 파일 이름을 설정해준다.

ysooj committed:
(add) 대화를 txt 파일로 저장하고, feedback에 불러오기
(fix) feedback_prompt, invoke 호출 방식, input_data 수정

ysooj committed:
fix/quiz의 질문 부분만 추출해서 feedback_prompt에 전달하기
- 정규식을 사용해 특정 패턴만 추출
- feedback_prompt의 '대화 기록'부분과 모델에 전달하는 데이터의 previous_conversation을 quiz_onlys로 수정

ysooj committed:
fix/리뷰에 제공한 코드를 바탕으로 코드 수정
- quiz_list 생성
- quiz의 질문 부분 내용을 추출해오는 곳 수정

ysooj committed:
Merge branch 'data/ML' of https://github.com/KooSuYeon/palddakpalddak

ysooj committed:
Merge branch 'main' of https://github.com/KooSuYeon/palddakpalddak

ysooj committed:
add/streamlit으로 ui 만들기
간단하게 ui를 구성해보았습니다.

ysooj commented:
@KooSuYeon 
간단하게 ui 구성해봤습니다. streamlit에서 구현해야 할 기능같은 게 있을까요? 어떤 걸 더 하면 좋을 지 모르겠습니다.

KooSuYeon commented:
@ysooj 
넵 수고하셨습니다!

우선은 간단하게 챗봇 화면 스크린샷 찍어서 올려주실 수 있으실까요?!
혹시 채팅 텍스트가 담길 message container도 구현이 되었을까요?!

ysooj commented:
@KooSuYeon 

![image](https://github.com/user-attachments/assets/d97cf76c-c2da-4ff2-9912-aa57933f1711)

@st.cache_data로 하면 새로고침했을 때 대화 내역이 사라지고, 
@st.cache_resource로 하면 새로고침해도 대화 내역이 그대로 남아있습니다!

UI 단에서 객관식 선택지를 제시하면 좋겠다고 하셔서, 우선 사이드바에 selectbox로 선택할 수 있게 했습니다.

KooSuYeon commented:
@ysooj 
오 넵 감사합니다!! 👍 수고 많으셨어요!
혹시 그러면 대화를 끝낸 경우에는 객관식 선택지 옆에 저장되는 걸까요?!

ysooj commented:
@KooSuYeon 
저희가 만든 챗봇이랑 streamlit을 합치려고 여러 가지 시도 중인데 잘 안 되네요ㅠㅠ 이미 기능은 저희가 다 구현해놨으니까 그걸 streamlit에 적용만 하면 될 것 같은데. 이것저것 해보겠습니다! 혹시 방법을 아신다면 가르쳐주시면 감사하겠습니다!

KooSuYeon commented:
@ysooj 
아 넵넵!
아마 간단하게는 
1. 저희가 만든 ipynb 파일 자체를 py 파일로 뺀 후
2. 챗봇의 quiz내용을 챗봇 응답 content로 넣어주고
3. 사용자의 input값을 갖고 
4. feedback chain에 넣어준 후 (invoke에)
5. 각 chain의 응답값을 챗봇의 content로 빼면 구동이 될 듯 합니다..!

혹시나 구동이 안되신다면은 모델이랑 프론트엔드 단에는 백엔드 (API)를 거쳐야 하기 떄문에 
구현사항이 약간은 달라질 수도 있어서 이따가 회의 시간에 같이 해결해보면 좋을듯 합니다!

ysooj commented:
@KooSuYeon 
오오 알겠습니다! 우선 사이드바에 채팅 기록을 남길 수 있는 방법을 계속 찾아보겠습니다!

ysooj committed:
DDAK-16 fix/사이드바에 대화내역 버튼 생성

ysooj reviewed (commented):
@KooSuYeon 
### 사이드바에 대화 내역 버튼 생성

- 대화 내역을 csv 파일로 저장하고 불러와서 사이드바에서 확인할 수 있게 했습니다.

- 버튼을 클릭해 해당 대화 내역 전체를 확인할 수 있습니다.

- 참고 코드

```python

CSV_FILE = "chat_history.csv"
try:
    chat_history_df = pd.read_csv(CSV_FILE)
except FileNotFoundError:
    chat_history_df = pd.DataFrame(columns=["ChatID", "Role", "Content"])
    
    def get_button_label(chat_df, chat_id):
    first_message = chat_df[(chat_df["ChatID"] == chat_id) & (chat_df["Role"] == "User")].iloc[0]["Content"]
    return f"Chat {chat_id[0:7]}: {' '.join(first_message.split()[:5])}..."


for chat_id in chat_history_df["ChatID"].unique():
    button_label = get_button_label(chat_history_df, chat_id)
    if st.sidebar.button(button_label):
        current_chat_id = chat_id
        loaded_chat = chat_history_df[chat_history_df["ChatID"] == chat_id]
        loaded_chat_string = "\n".join(f"{row['Role']}: {row['Content']}" for _, row in loaded_chat.iterrows())
        st.text_area("Chat History", value=loaded_chat_string, height=300)

```

[출처][(https://discuss.streamlit.io/t/how-to-create-a-chat-history-on-the-side-bar-just-like-chatgpt/59492/4)](https://discuss.streamlit.io/t/how-to-create-a-chat-history-on-the-side-bar-just-like-chatgpt/59492/4)


-------------------------------

### 트러블슈팅

참고 코드에서는 ChatID와 Content(대화 내역 전체)를 기반으로 버튼을 생성했습니다. Content 대신 가장 최근에 사용자가 입력한 메시지가 보이게 하기 위해서 get_button_label 함수를 수정했습니다. 그 과정에서 IndexError: single positional indexer is out-of-bounds 오류가 발생했습니다.



참고 코드에서는 chat_df[(chat_df["ChatID"] == chat_id) & (chat_df["Role"] == "User")].iloc[-1]["Content"]를 사용했습니다.

그런데 이 경우에는, User 메시지가 없을 경우 IndexError가 발생한다고 합니다.



"User" 대신 "user"로 수정하고 예외처리를 해주면, User 메시지가 없는 경우 IndexError는 발생하지 않고, "No User message found"라는 메시지가 표시됩니다.
---------------

ysooj commented:
@KooSuYeon 
수연님께서 알려주신 대로 코드를 수정해보고 있는데, 챗봇이 잘 실행되지 않습니다ㅠㅠ
계속 여러 방법으로 시도해보겠습니다!
우선적으로 사이드바에 대화 내역을 확인할 수 있는 부분을 추가했습니다.

HSgoon committed:
Add files via upload

#start_button 구현 중
#진행 바 구현 중
#side bar  디폴트 페이지 구성 중
#side bar  옵션 선택 시 알맞은 과목만 출력 구현코드 만드는 중

HSgoon commented:
@KooSuYeon @ysooj 
![스크린샷 2024-11-28 133036](https://github.com/user-attachments/assets/61aa8646-4641-4180-8530-1283d8624477)
![스크린샷 2024-11-28 133059](https://github.com/user-attachments/assets/e2c0ed45-6493-45f7-919d-68da7aed2284)


아직 구현 중인 세팅이 많습니다.
4번까지의 객관식 선택지가 있는 문제를 구현하는 퀴즈 페이지입니다.

ysooj committed:
DDAK-3 사이드바에 목차와 대화내역 구현

ysooj committed:
Merge branch 'streamlit' of https://github.com/KooSuYeon/palddakpalddak into streamlit

ysooj committed:
DDAK-3 fix/대화내역 버튼 수정

ysooj committed:
DDAK-16 fix/목차 선택 selectbox 만들기

ysooj commented:
@KooSuYeon 
목차 선택하는 부분까지 구현 완료했습니다. 확인해주시면 감사하겠습니다.
-------------------
스크린샷은 아래와 같습니다.

### 대화 시작 전(대화 내역 csv 파일 생성 전)
![image](https://github.com/user-attachments/assets/a2ddd4cc-72a8-48f1-a440-b89df4017049)

### 대화 진행
![image](https://github.com/user-attachments/assets/6d69759d-e716-475f-8d73-22a8ba621b7b)

### 대화 내역 버튼 선택 시
![image](https://github.com/user-attachments/assets/70f0ff7c-69af-4460-ace0-46f8460e748d)
![image](https://github.com/user-attachments/assets/f916b645-01f2-4235-bdd6-dcfa13d3424c)
-------------------------
- 대화 내역은 새로고침이나 페이지를 나갔다가 들어와서 대화를 시작할 때마다 생성됩니다.
- 주제와 교재를 선택할 수 있습니다.
  선택한 주제나 교재를 대화 중간에 바꾸더라도 페이지가 이동하거나 대화 내역이 바뀌지는 않습니다.

ysooj committed:
DDAK-16 OPENSOURCE 목차 추가

ysooj committed:
DDAK-16 fix/mapping_data 추가
